README
                                                        ____     _        
  __                 _      _                          / ___|   / |       
 / _|  _        _   | \    / |  _                     | ||  \/\/  |       
| |_  |_| |  | |_   |  \/\/  | |_| |  |               | |___      |       
 \__| | |  \/  |_   |___/\___| | | |_ |_               \____|/\___|       
                                             ()  ^                        
.-''-.-''-.-''-.-''-.-''-.-''-.-''-.-''-._ /''''\| ____.-''-.-''-.-''-.  
'-..-'-..-'-..-'-..-'-..-'-..-'-..-'-..-'-. /""\ | _..-'-..-'-..-'-..-'
                                                                          
  [ Colored ASCII Visual Editor Without Any Loathsome Limitations ]

!!! WARNING !!!
This project is no logner actively developed or maintained.

This is a development version. Author is not responsible for any damage caused by this software. Use at your own risk!

This software is distributed under the terms of GNU General Public License v.2.0 (see the "LICENSE" file).
Please also check the "BUGS" file.

0. INTRODUCTION

CAVE WALL (Colored ASCII Visual Editor Without Any Loathsome Limitations) is an ASCII Art editor for Linux, somewhat similiar to such programs as JavE, Aewan, DuhDraw or EDASCII.  It is distributed on the terms of GNU General Public License. At the time it is in early development stage, but usable. It's written in C and the interface is based on the ncurses library.

It has not been tested on systems other than Linux, if you would like to do it, then - do it, and tell me about the results. I'd be very grateful if you did. It also has not been tested for compatibility with other (not ncurses) implementations of the curses library.

The project, originally named CAVE, was started in 2002 by Michal Miszewski, then left for a while, and revived in 2004 by me. The name was changed because it had already been reserved by some other project. CAVE WALL is my first bigger project and I took it up mostly in order to learn a bit of C programming.

1. INSTALLATION

First make sure you have some recent version of ncurses installed. After unpacking the cavewall-x.x.tar.gz tarball check and adjust the "config.h" file to your needs, then  just run "make" in the unpacked directory and the program should compile. If nothing went wrong you can run "./cave" to start the program. You may also copy the "cave" file to /usr/local/bin or wherever you want to make it available systemwide.

2. USING CAVE WALL

You can start CAVE WALL either by typing "cave" or "cave [FILENAME]". In the first case program starts and shows a banner (or a blanks screen if it was so set during setup). In the second one it loads the file specified in the optional FILENAME parameter. If the file does not exist it will not be created, but the filename will be set as default filename for the export option.

Once you have started CW you should see black screen with a blue line at the bottom. The line is called status bar. It shows (in order from left to right):
- red B letter if in blockmode (see BLOCK OPERATIONS section)
- cursor's x (column it is in)
- cursor's y (row)
- character the cursor is placed over
- current foreground color
- current background color
- current attributes - this displays the first character of name of each attribute that is currently set. The character is displayed using the attribute it is connected with (so, when "bold" is on, a 'b' character will be displayed in bold; when "underline" is on, an underlined 'u' will be displayed, etc.).
- current cursor step mode
- last pressed key's keynumber
- a demonstration of how will printed characters look with current foreground color, background color and attributes set

The black part of the screen shows part of the edit pad (because usually it is larger than your terminal).

Now move the cursor using the cursor keys of your keyboard. Press a letter to print thr letter on the screen. Move the cursor to the edge of the screen and further to scroll through the edit pad.  Pressing PgUp, PgDown, End and Home keys moves the cursor to the edge of the screen. If it is already there, it scrolls the edit pad by one screen. CAVE WALL supports images up to 1000x1000 characters, but remember that editing large images consumes much memory. With current systems that's usually not a big problem, but if you have 16M of RAM or so, then watch out! It's also nice to know that the pad is not 1000x1000 at the beginning, but it is resized when you scroll it. It is not shrinked though, when you scroll back, even if no characters were printed. Pressing space does not earse the character under cursor - Delete and Backspace keys do so. Space just changes it's color, background color and attributes to currently set ones. Enter clears to the end of the current line.

Some important operations are performed using the function keys. The default key bindings (not changed in "config.h" during installation) are:

F1 - show menu
F2 - export image
F3 - import image
F4 - block operations
F5 - show color/background selection dialog
F6 - show attribute selection dialog
F7 - show cursor step mode selection dialog
F8 - NOT USED
F9 - NoT USED
F10 - paste from clipboard
F11 - undo
F12 - clear screen

Here's a more detailed info on each of these operations:

F1 - show menu - shows a menu, you can move using the arrow keys or number keys, choose menu positions by hitting enter. Here's a short description of menu positions:
	Help - Actually it just displays the function keys list that is above.
	Select fg/bg color - shows a foreground/background color selection dialog - it may be faster than setting them using F5/F6 keys
	Select attributes - shows a attribute  selection dialog - it may be faster than setting them using F7 key
	Select step mode - shows a step mode selection dialog - it may be faster than it using F8 key
	Insert FIGletters - available if you set it in config.h and have FIGlet (http://www.figlet.org) installed - shows a dialog that's used for inserting FIGlet output in current cursor position (THIS DOES NOT WORK YET)
	Exit menu - exits menu :) 
	Quit CAVE WALL - quits program - WARNING: without any "are you really really sure"-like questions and without possibility to save your work. So - first save, then quit!
F2 - export image - three formats are supported: Text, ANSI and HTML. Choose the format and then enter filename. Tilde ( ~ ) is NOT allowed. Export automatically deletes empty spaces at the end of lines and empty lines at the end of file. HTML export uses CSS. The created HTML file will link to stylesheet cavewall.css, which is included in the cavewall.XXXX.tar.gz package.
F3 - import image - just enter filename (tilde IS allowed, wildcards NOT allowed), CAVE WALL will automatically interprete any ANSI codes
F4 - block operations - that's a more advanced one. Check appropriate section below.
F5 - Shows color selection dialog. Use arrow keys to move around. Pressing Enter will choose the currently selected fg/bg combination. New drawn characters will have this color.
F6 - Shows character attribute selection dialog. Use arrow or number keys to move around, press Enter to select/deselect an attribute. Selected attributes have a "*" next to their name. Selecting first attribute ("normal") deselects all attributes. Choose the last position ("Exit") to exit dialog. Note that attributes (and colors too, but maybe in a less noticeable way) are displayed in different ways on different terminals. So don't panic when your blinking character does not blink.
F7 - Shows cursor step mode selection dialog.
     What are the step modes? They determine where will the cursor move after a character is drawn. This can be:
	- normal - does not move at all
	- step_fwd - one character forward (right)
	- step_bwd - one character backward (left)
	- step_up - one charatcer up
	- step_dn - one character down
	- step_ul - one up, one left
	- step_ur - one up, one right
	- step_dl - one down, one left
	- step_dr - one down, one right
	- step_rnd - one step in random direction (including diagonal directions), so it can be any of the above excluding normal. This can be switched off in "config.h" during installation.
	(or: change cursor step mode - scrolls through the step modes list).
F8 - NOTHING AS FOR NOW
F9 - NOTHING AS FOR NOW
F10 - paste from clipboard - pastes a previously copied selection (see BLOCK OPERATIONS section). The upper left-hand corner of the selection is placed at the current cursor position. You can paste with transparency - characters which are spaces with default colors and attributes are treated as transparent.
F11 - undo - undoes last operation. And only the last one. Operations that can be undid are: character drawing and color/attribute changes, block operations excluding copy, pasting, undoing. Since undo is an undoable operation, pressing this again is redo.
F12 - clear screen - clears the edit pad completly and resizes it to one screen size. This CAN NOT be undid.

3. BLOCK OPERATIONS

Block operations are operations on rectangular selections. Some of them are useful and can speed up your work. Some probably are not.

Here's how to use them in general:

Press F4. A red "B" will appear on the left of the status bar. You are now in block selection mode. Pressing F4 again without moving the cursor or after moving it and then coming back to the same place will disable it. But don't disable it. Now, in block mode, move cursor to some other place and press F4 again. Notice how your selection is highlighted (if you didn't turn it off in config.h). This highlighting may seem useful (as for me, I can't say it really is). Now, a rectangle with two opposite corners in the points you choose by pressing F4 is selected. This rectangle is your block. A menu will appear on the status bar. Read it. It is the list of available block operations. Press the [G] key to see more of them. A full list of block operations is below.  Try some option by pressing adequate keys ([A],[B],[C], etc.). After using one of them (if it was not the "e.DelBlock" one) red "B" will remain on the status bar, but something will probably happen to your image. Press F4 again - the menu will appear. Press [E] to remove the block. The red "B" will disappear. Then press F4 again if you want to start selecting a new block. That's it. Notice you can draw characters etc. while in block mode. 

Here's the block operations list:

Page 1:
    a.Copy - copies the selection to clipboard for pasting in the future.
    b.Cut - same as above, but clears the selection after copying it
    c.Clear - clears the selection
    d.Fill - fills the selection either with entered string or with the content of clipboard. In the first case there are four ways of filling. Here's an example:

 We want to fill a 7x7 selection with the "foobar" string. Here are the results for different modes:
	
 foobarf  Vertical Lines
 foobarf
 foobarf
 foobarf
 foobarf
 foobarf
 foobarf

 fffffff  Horiznotal Lines
 ooooooo
 ooooooo
 bbbbbbb
 aaaaaaa
 rrrrrrr
 fffffff

 foafoaf  Vertically
 obrobro
 oafoafo
 brobrob
 afoafoa
 robrobr
 foafoaf
 obrobro

 foobarf  Horizontally
 oobarfo
 obarfoo
 barfoob
 arfooba
 rfoobar
 foobarf

    e.DelBlock - removes the curren selection, so that you are no longer in block mode and can select a new one.
    f.Cancel - turn the menu off, but stay in block mode
    g.More--> - show next page

Page 2:
    a.Colorize - gives currently set attribute and colors to all characters in the selection
    b.Rotate - rotates the selection by 90, 180 or 270 degrees. The upper left-hand corner of image after rotation is placed in the upper left-hand corner of the selection.
    c.MirrorH - mirrors the selection horizontally, with optional character conversion. Some characters are "mirrorable", that means there are some other characters that (more or less) look like them flipped. You can use the conversion or not. Here's a list of convertable characters: \ <-> /, ` <-> ,, . <-> ', _ <-> ", ^ <-> v, M <-> W, m <-> w I know you may not like this list, the conversion table should be placed in a separate file soon.
    d.MirrorV - mirrors the selection vertically, with optionall character conversion. Just like above, but for other characters: all types of brackets, b <-> d, p <-> q, > <-> <, / <-> \     
    e.DrawBox - draws a 1-character wide box around the selection. You can specify a different character sequence for each wall. Corners are not drawn.
    f.Cancel - just like cancel in page one
    g.Previous--> - previous page
    
4. CONTACT

You can contact me by email: kukimik@poczta.onet.pl
CAVE WALL Homepage: http://www-users.mat.umk.pl/~mckuk/cavewall/index.html

5. HAVE FUN!


Changelog
2004-06-30 Kuki <kukimik@poczta.onet.pl>

    * mirrorable characters conversion in mirror block operations
    * shortened too long text on status bar in block operations mode
    
2004-07-01 Kuki <kukimik@poczta.onet.pl>

    * fixed a bug in ANSI import that set color and background to white and black when ANSI code set only new attribute
    * more ANSI codes support in import.c (ESC[2J, ESC[K and cursor movement codes)
    
2004-07-08 Kuki <kukimik@poczta.onet.pl>

    * finally added support for large pics (up to 1000x1000, more would screw up status bar coordinates display) - may be buggy and is not optimized
    * fixed a bug in rotate undo
    * modified the logo a bit
    
2004-07-11 Kuki <kukimik@poczta.onet.pl>

    * removed a bug saving undo even if no block opearation was done
    * removed some bugs in large pics support
    * cosmetic changes to README
    
2004-07-14 Kuki <kukimik@poczta.onet.pl>

    * 0.2 release
    
2004-07-25 Kuki <kukimik@poczta.onet.pl>

    * simple parameter support
    * shows usage note when too many parameters specified
    * key changes - paste F10, undo F11
    * removed a bug in import preventing import of files larger than current editpad size
    * found out the curses wresize() function - change_editp_size() rewritten, now works a whole lot faster

2004-07-26 Kuki <kukimik@poczta.onet.pl>

    * parameters now work ok
    * redo support
    * a nice, 10kB README

2004-07-28 Kuki <kukimik@poczta.onet.pl>

    * removed some bugs in undo saving, including removing undo information before import, even if nothing was imported and deleting pads that did not exist 
    * 0.21 release
    
2004-08-03 Kuki <kukimik@poczta.onet.pl>

    * move cursor to (0,0) after importing (to avoid having parts of previous picture remaining on screen)

2004-08-12 Kuki <kukimik@poczta.onet.pl>

    * added transparent paste
    * import functions, cut, block_clear and rmch now use default colors and attributes
    * set default background to editp with wsetbkgd
    * protected save_undo from saving undo outside the editp (occured when pasting at the edge of a maximum-sized editp)
    * removed a yes=no bug in the "paste image larger than editp" question
    * enter now officialy clears to eol
    
2004-08-21 Kuki <kukimik@poczta.onet.pl>

    * block selection highlighting - not really optimized but seems to work well
    * removed a bug - filling block with empty string no longer fills it with crap
    * documented new features in README
    * 0.25 release

2004-08-22 Kuki <kukimik@poczta.onet.pl>

    * removed a bug - box with empty-string walls no longer draws crap

2004-08-24 Kuki <kukimik@poczta.onet.pl>

    * some minor changes in block_drawbox

2004-12-07 Kuki <kukimik@poczta.onet.pl>

    * basic HTML export

2005-01-14 Kuki <kukimik@poczta.onet.pl>

    * reverse support in HTML export

2006-01-14 Kuki <kukimik@poczta.onet.pl>

    * simple menu under F1 (instead of help screen)
    * moved help to menu and modified help screen a bit
    * "quit program" position in menu (at last some nice way to exit CAVE WALL :) ) 
    * "exit menu" position in menu
    * some changes in README regarding the menu feature
    
2006-05-21 Kuki <kukimik@poczta.onet.pl>

    * changes in menu layout
    * color selection as menu position (to be moved to a dialog box)

2006-05-28 Kuki <kukimik@poczta.onet.pl>

    * killed highlighting bug

2006-06-12 Kuki <kukimik@poczta.onet.pl>

    * color selection in dialog box

2006-07-05 Kuki <kukimik@poczta.onet.pl>

    * little NODIALOG issues
    * fixed color selection dialog bug
    
2006-07-06 Kuki <kukimik@poczta.onet.pl>

    * unused function keys do not produce crap on screen
    * step mode selection dialog
    * 0.26 release    
    
2006-07-08 Kuki <kukimik@poczta.onet.pl>

    * new menu and dialogs handling
    * fixed HTML export bug
    * 0.265 release
    * CSS HTML export
    * little README changes (CSS export, new website URL)
    * removed old HTML export code

2006-07-10 Kuki <kukimik@poczta.onet.pl>

    * multiattribute support
    * fixed dialogs bug (cursor moving to row <0)
    * README changes
    
2006-07-12 Kuki <kukimik@poczta.onet.pl>

    * fixed color dialog bug
    * PgUp, PgDn support in dialogs (and Home/End in color selection dialog)
    * changed function keys order, now F8 and F9 are not used
    * 0.27 release

menu.c
/*
 *  menu.c
 *
 *  CAVE WALL - Colored Ascii Visual Editor Without Any Loathsome Limitations
 *  General routines
 *
 *  Copyright (C) 2006 Kuki <kukimik@poczta.onet.pl>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License version 2 as
 *  published by the Free Software Foundation.
 */

#include <curses.h>
#include <stdlib.h>
#include "menu.h"
#include "block.h"
#include "cave.h"
#include "conffile.h"
#include "color.h"
#include "config.h"
#include "cursor.h"
#include "status.h"
#include "import.h"
#include "cursor.h"
#include "export.h"
#ifdef BANNER
#include "banner.h"
#endif


extern struct config
{
  int defcolor, defbgnd, defattr, defmode, fromzero,
  key_up, key_down, key_left, key_right,
  key_leftborder, key_rightborder, key_upborder, key_downborder,
  key_delch, key_backspc, key_addattribs,
  key_export, key_import, key_color, key_bgnd, key_attrib, key_mode, key_clear, key_block, key_paste, key_undo, key_menu, key_unknown;
} conf;

// ===== GLOBAL VARIABLES =====

WINDOW *menuw, *dialogw;


// ==== FUNCTIONS ====

//
/* shows help screen */
//
void help_screen(void)
{  
wclear(menuw);
attron(A_BOLD);
mvwprintw(menuw,1,xsize/2-2,"HELP");
attroff(A_BOLD);
mvwprintw(menuw,3,2,"F1 - show menu");
mvwprintw(menuw,4,2,"F2 - export");
mvwprintw(menuw,5,2,"F3 - import");
mvwprintw(menuw,6,2,"F4 - block operations");
mvwprintw(menuw,7,2,"F5 - select foreground/background color");
mvwprintw(menuw,8,2,"F6 - select attributes");
mvwprintw(menuw,9,2,"F7 - select cursor step mode");
mvwprintw(menuw,12,2,"F10 - paste from clipboard");
mvwprintw(menuw,13,2,"F11 - undo");
mvwprintw(menuw,14,2,"F12 - clear screen");
mvwprintw(menuw,16,(xsize-2)/2-13,"Press any key to continue.");
wrefresh(menuw);
getch();
return;
}

//
/* menu and dialogs handling */
//
char menu_handle(char last, char choice, const short int x0, const unsigned int y0, WINDOW *dialogwindow)
{
short int x=choice+x0;
int key;
mvwprintw(dialogwindow,x,y0,"->");	
wrefresh(dialogwindow);
do
{
key=getch();

if (key==conf.key_up)
{	
    mvwprintw(dialogwindow,x,y0,"  ");
    x-=1;
    if (x<=x0) x=x0;
    mvwprintw(dialogwindow,x,y0,"->");	
    wrefresh(dialogwindow);
}	

if (key==conf.key_down)
{	
    mvwprintw(dialogwindow,x,y0,"  ");
    x+=1;
    if (x>=last+x0) x=x0+last-1;
    mvwprintw(dialogwindow,x,y0,"->");	
    wrefresh(dialogwindow);
}

if (key==conf.key_upborder)
{	
    mvwprintw(dialogwindow,x,y0,"  ");
    x=x0;
    mvwprintw(dialogwindow,x,y0,"->");	
    wrefresh(dialogwindow);
}

if (key==conf.key_downborder)
{	
    mvwprintw(dialogwindow,x,y0,"  ");
    x=x0+last-1;
    mvwprintw(dialogwindow,x,y0,"->");	
    wrefresh(dialogwindow);
}

if (last<=10)
{
if ((key<'0'+last) && (key>='0'))
{
mvwprintw(dialogwindow,x,y0,"  ");
x=x0+key-'0';
mvwprintw(dialogwindow,x,y0,"->");	
wrefresh(dialogwindow);
}
}

if (key==10) // enter
{
    choice=x-x0;
}

} while (key!=10); // inner loop - for moving around in the menu and choosing menu positions


return choice;
}

//
/* attribute selection dialog */
//
void attr_dialog(void)
{
if ((dialogw = newwin(NUMATTRS+1,15, 0, 0)) == NULL) printf("error creating dialog window\n");
int i;
unsigned short choice=0;
wclear(dialogw);

if (NUMATTRS<=10)
{
    for (i=0; i<NUMATTRS; i++)
    {
    mvwaddch(dialogw,i,2,'0'+i);
    mvwaddch(dialogw,i,3,'.');
    mvwprintw(dialogw,i,4,"%c%s", ( attribute & avals[i] ? '*' : ' '), attrs[i]);
    }
    mvwaddch(dialogw,i,2,'0'+i);
    mvwaddch(dialogw,i,3,'.');
    mvwprintw(dialogw,i,5,"Exit" );
}
else
{
    for (i=0; i<NUMATTRS; i++)
    {
    mvwprintw(dialogw,i,4,"%c%s", ( attribute & avals[i] ? '*' : ' '), attrs[i]);
    }
    mvwprintw(dialogw,i,5,"Exit" );
}

// Allow user to move around
do
{
    choice=menu_handle(NUMATTRS+1,choice,0,0,dialogw);
    if (choice==0) 
    {
    attribute=A_NORMAL;
    for (i=0; i<NUMATTRS; i++) mvwaddch(dialogw,i,4,' ');    
    update_status();
    }

    else if (!(choice==NUMATTRS))
    {
    if (attribute & avals[choice]) {attribute = attribute & ~(avals[choice]); mvwaddch(dialogw,choice,4,' ');}
    else { attribute = attribute | avals[choice]; mvwaddch(dialogw,choice,4,'*'); }
    update_status();
    }
} while (choice!=NUMATTRS);

touchwin(editp);
editprefresh();
delwin(dialogw);
return;
}

//
/* color selection dialog */
//
void color_dialog(void)
{  
if ((dialogw = newwin(11,29, 0, 0)) == NULL) printf("error creating dialog window\n");

// draw the color box
int i, j;
wclear(dialogw);
mvwprintw(dialogw,0,15,"FG");
mvwprintw(dialogw,5,0,"B");
mvwprintw(dialogw,6,0,"G");
mvwprintw(dialogw,1,4," 0  1  2  3  4  5  6  7");
for (i=0; i<8; i++)
{
    for (j=0; j<8; j++)
    {
    mvwaddch(dialogw,i+2,3*j+5,'#' | COLOR_PAIR(find_pair(i,j)) | attribute );
    }
mvwaddch(dialogw,i+2,2,'0'+i);
}
wrefresh(dialogw);


// Allow user to move around
char y=(color+1)*3;
char x=background+2;
int key;
mvwprintw(dialogw,x,y,"->");	
wrefresh(dialogw);
do
{
key=getch();

if (key==conf.key_left)
{
    mvwprintw(dialogw,x,y,"  ");
    y-=3;
    if (y<3) y=3;
    mvwprintw(dialogw,x,y,"->");	
    wrefresh(dialogw);
}

if (key==conf.key_right)
{
    mvwprintw(dialogw,x,y,"  ");
    y+=3;
    if (y>24) y=24;
    mvwprintw(dialogw,x,y,"->");	
    wrefresh(dialogw);
}

if (key==conf.key_up)
{	
    mvwprintw(dialogw,x,y,"  ");
    x-=1;
    if (x<2) x=2;
    mvwprintw(dialogw,x,y,"->");	
    wrefresh(dialogw);
}	

if (key==conf.key_down)
{	
    mvwprintw(dialogw,x,y,"  ");
    x+=1;
    if (x>9) x=9;
    mvwprintw(dialogw,x,y,"->");	
    wrefresh(dialogw);
}	

if (key==conf.key_upborder)
{	
    mvwprintw(dialogw,x,y,"  ");
    x=2;
    mvwprintw(dialogw,x,y,"->");	
    wrefresh(dialogw);
}

if (key==conf.key_downborder)
{	
    mvwprintw(dialogw,x,y,"  ");
    x=9;
    mvwprintw(dialogw,x,y,"->");	
    wrefresh(dialogw);
}

if (key==conf.key_leftborder)
{	
    mvwprintw(dialogw,x,y,"  ");
    y=3;
    mvwprintw(dialogw,x,y,"->");	
    wrefresh(dialogw);
}

if (key==conf.key_rightborder)
{	
    mvwprintw(dialogw,x,y,"  ");
    y=24;
    mvwprintw(dialogw,x,y,"->");	
    wrefresh(dialogw);
}

if (key==10) // enter
{
    color=y/3-1;
    background=x-2;
    pair=find_pair(y/3-1,x-2);
    update_status();
}

} while (key!=10); // inner loop - for moving around in the menu and choosing menu positions
delwin(dialogw);

touchwin(editp);
editprefresh();
return;
}

//
/* step mode selection dialog */
//
void step_dialog(void)
{
if ((dialogw = newwin(NUMMODES,14, 0, 0)) == NULL) printf("error creating dialog window\n");
int i;
wclear(dialogw);

if (NUMMODES<=10)
{
    for (i=0; i<NUMMODES; i++)
    {
    mvwaddch(dialogw,i,2,'0'+i);
    mvwaddch(dialogw,i,3,'.');
    mvwprintw(dialogw,i,4,modes[i]);
    }
}
else
{
    for (i=0; i<NUMMODES; i++)
    {
    mvwprintw(dialogw,i,2,modes[i]);
    }
}

// Allow user to move around
mode=menu_handle(NUMMODES,mode,0,0,dialogw);

update_status();
touchwin(editp);
editprefresh();
delwin(dialogw);
return;
}


//
/* show usable menu window */
//
void menu_screen(void)
{
bool exit_menu=false;
char choice=0;
const int MAXPOSITION=7;
if ((menuw = newwin(ysize-1,xsize, 0, 0)) == NULL) printf("error creating menu window\n");

do
{
wclear(menuw);
attron(A_BOLD);
mvwprintw(menuw,1,xsize/2-2,"MENU");
mvwprintw(menuw,2,xsize/2-3,"------");
attroff(A_BOLD);
mvwprintw(menuw,3,3,"0. Help");
mvwprintw(menuw,4,3,"1. Select fg/bg color");
mvwprintw(menuw,5,3,"2. Select attributes");
mvwprintw(menuw,6,3,"3. Select step mode");
#ifdef FIGLET
mvwprintw(menuw,7,3,"4. Insert FIGletters");
#else
mvwprintw(menuw,7,3,"4. Insert FIGletters (OFF)");
#endif
mvwprintw(menuw,8,3,"5. Exit menu");
mvwprintw(menuw,9,3,"6. Quit CAVE WALL");
wrefresh(menuw);

choice=menu_handle(MAXPOSITION,choice,3,1,menuw);

    switch (choice)
    {    
    case 0: // show help screen
    help_screen();
    break; 
    
    case 1: // color selection dialog
    color_dialog();
    break;

    case 2: // attributes selection dialog
    attr_dialog();
    break;

    case 3: // step mode selection dialog
    step_dialog();
    break;

    case 5: // exit menu - handled in the while loop
    exit_menu=true;
    break;
        
    case 6: // maxposition - quit program
    exit_program=true;
    break;
    }

} while (exit_menu==false && exit_program==false); // outer loop - shows menu

delwin(menuw);

if (exit_program==false) // when exiting menu only
{
touchwin(editp);
editprefresh();
}

return;
}


